src/register-service-worker.js

```diff
@@ -0,0 +1,162 @@
+/* eslint-env browser */
+
+(function() {
+  var constants = require('./lib/constants.js');
+
+  var swScript = document.currentScript.dataset.serviceWorker;
+  var manifestAttribute = document.documentElement.getAttribute('manifest');
+
+  if (swScript && manifestAttribute && 'serviceWorker' in navigator) {
+    var manifestUrl = (new URL(manifestAttribute, location.href)).href;
+
+    var idb = require('idb');
+
+    idb.open(constants.DB_NAME, constants.DB_VERSION, function(upgradeDB) {
+      if (upgradeDB.oldVersion === 0) {
+        Object.keys(constants.OBJECT_STORES).forEach(function(objectStore) {
+          upgradeDB.createObjectStore(constants.OBJECT_STORES[objectStore]);
+        });
+      }
+    }).then(function(db) {
+      return Promise.all([
+        updateManifestAssociationForCurrentPage(db, manifestUrl),
+        handlePossibleManifestUpdate(db, manifestUrl)
+      ]);
+    }).then(function() {
+      return navigator.serviceWorker.register(swScript);
+    });
+  }
+
+  /**
+   * Caches the Responses for one or more URLs, using the Cache Storage API.
+   *
+   * @param {Array.<String>} urls
+   * @returns {Promise.<T>}
+   */
+  function addToCache(urls) {
+    return caches.open(constants.CACHE_NAME).then(function(cache) {
```

Comment:
> In practice, I think I'm going to have to hash the manifest contents to assign them version ids

Agreed, this is what I did in the template https://github.com/jakearchibald/appcache2serviceworker/blob/master/templates/serviceworker.hbs#L5

I don't think the `unload` handler is safe enough here, given cache-deleting is async. 

I think I was wrong with `postMessage`, as I thought the client would be the thing that decides which cache it should use, but that's too late. I think the SW should orchestrate it. Lack of something like https://github.com/slightlyoff/ServiceWorker/issues/817 is making this tough.

How about this:

1. If a fetch as no clientId, it will use the newest version of the relevant manifest & cache.
1. If a fetch has a clientId, look up which manifest & cache it should use in IDB (client-to-cache)
1. If no cache is listed for that clientId, assign it the newest version of the relevant manifest & cache and record this in IDB client-to-cache
1. Return response, but then in the background:
1. If the current fetch is a navigation, get all the clients for the origin, remove client-to-cache entries for clients that no longer exist
1. Remove caches & manifests that have been superseded & no longer have clients using them

Since the client-to-cache map is established at the first subresource request, there's a race where the client request could end up with a different cache than its subresources. This would happen if an update completes between the page request and the first subresource request. This would be extremely unlikely, but it's a shame that it's possible. https://github.com/slightlyoff/ServiceWorker/issues/817 would have given us what we needed,